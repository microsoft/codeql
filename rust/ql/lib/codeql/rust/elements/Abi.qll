// generated by codegen, do not edit
/**
 * This module provides the public class `Abi`.
 */

private import internal.AbiImpl
import codeql.rust.elements.AstNode

/**
 * An ABI specification for an extern function or block.
 *
 * For example:
 * ```rust
 * extern "C" fn foo() {}
 * //     ^^^
 * ```
 */
final class Abi = Impl::Abi;
