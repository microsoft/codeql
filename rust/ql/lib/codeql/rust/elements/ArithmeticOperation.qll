/**
 * Provides classes for arithmetic operations.
 */

private import codeql.rust.elements.BinaryExpr
private import codeql.rust.elements.PrefixExpr
private import codeql.rust.elements.Operation
private import codeql.rust.elements.AssignmentOperation

/**
 * An arithmetic operation, such as `+`, `*=`, or `-`.
 */
abstract private class ArithmeticOperationImpl extends Operation { }

final class ArithmeticOperation = ArithmeticOperationImpl;

/**
 * A binary arithmetic operation, such as `+` or `*`.
 */
final class BinaryArithmeticOperation extends BinaryExpr, ArithmeticOperationImpl {
  BinaryArithmeticOperation() { this.getOperatorName() = ["+", "-", "*", "/", "%"] }
}

/**
 * An arithmetic assignment operation, such as `+=` or `*=`.
 */
final class AssignArithmeticOperation extends BinaryExpr, ArithmeticOperationImpl,
  AssignmentOperation
{
  AssignArithmeticOperation() { this.getOperatorName() = ["+=", "-=", "*=", "/=", "%="] }
}

/**
 * A prefix arithmetic operation, such as `-`.
 */
final class PrefixArithmeticOperation extends PrefixExpr, ArithmeticOperationImpl {
  PrefixArithmeticOperation() { this.getOperatorName() = "-" }
}
