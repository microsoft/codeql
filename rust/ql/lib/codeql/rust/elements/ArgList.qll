// generated by codegen, do not edit
/**
 * This module provides the public class `ArgList`.
 */

private import internal.ArgListImpl
import codeql.rust.elements.AstNode
import codeql.rust.elements.Expr

/**
 * A list of arguments in a function or method call.
 *
 * For example:
 * ```rust
 * foo(1, 2, 3);
 * // ^^^^^^^^^
 * ```
 */
final class ArgList = Impl::ArgList;
