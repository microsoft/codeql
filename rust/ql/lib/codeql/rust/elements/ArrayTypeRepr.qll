// generated by codegen, do not edit
/**
 * This module provides the public class `ArrayTypeRepr`.
 */

private import internal.ArrayTypeReprImpl
import codeql.rust.elements.ConstArg
import codeql.rust.elements.TypeRepr

/**
 * An array type representation.
 *
 * For example:
 * ```rust
 * let arr: [i32; 4];
 * //       ^^^^^^^^
 * ```
 */
final class ArrayTypeRepr = Impl::ArrayTypeRepr;
