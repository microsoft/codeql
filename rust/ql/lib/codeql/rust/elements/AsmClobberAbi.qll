// generated by codegen, do not edit
/**
 * This module provides the public class `AsmClobberAbi`.
 */

private import internal.AsmClobberAbiImpl
import codeql.rust.elements.AsmPiece

/**
 * A clobbered ABI in an inline assembly block.
 *
 * For example:
 * ```rust
 * use core::arch::asm;
 * asm!("", clobber_abi("C"));
 * //       ^^^^^^^^^^^^^^^^
 * ```
 */
final class AsmClobberAbi = Impl::AsmClobberAbi;
