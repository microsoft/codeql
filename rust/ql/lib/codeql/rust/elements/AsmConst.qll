// generated by codegen, do not edit
/**
 * This module provides the public class `AsmConst`.
 */

private import internal.AsmConstImpl
import codeql.rust.elements.AsmOperand
import codeql.rust.elements.Expr

/**
 * A constant operand in an inline assembly block.
 *
 * For example:
 * ```rust
 * use core::arch::asm;
 * asm!("mov eax, {const}", const 42);
 * //                       ^^^^^^^
 * ```
 */
final class AsmConst = Impl::AsmConst;
