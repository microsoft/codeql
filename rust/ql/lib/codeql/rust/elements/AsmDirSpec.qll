// generated by codegen, do not edit
/**
 * This module provides the public class `AsmDirSpec`.
 */

private import internal.AsmDirSpecImpl
import codeql.rust.elements.AstNode

/**
 * An inline assembly direction specifier.
 *
 * For example:
 * ```rust
 * use core::arch::asm;
 * asm!("mov {input:x}, {input:x}", output = out(reg) x, input = in(reg) y);
 * //                                        ^^^                 ^^
 * ```
 */
final class AsmDirSpec = Impl::AsmDirSpec;
