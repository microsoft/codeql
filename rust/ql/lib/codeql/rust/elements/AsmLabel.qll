// generated by codegen, do not edit
/**
 * This module provides the public class `AsmLabel`.
 */

private import internal.AsmLabelImpl
import codeql.rust.elements.AsmOperand
import codeql.rust.elements.BlockExpr

/**
 * A label in an inline assembly block.
 *
 * For example:
 * ```rust
 * use core::arch::asm;
 * asm!(
 *     "jmp {}",
 *     label { println!("Jumped from asm!"); }
 * //  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 * );
 * ```
 */
final class AsmLabel = Impl::AsmLabel;
