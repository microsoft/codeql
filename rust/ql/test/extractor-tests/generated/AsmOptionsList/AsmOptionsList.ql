// generated by codegen, do not edit
import codeql.rust.elements
import TestUtils

from AsmOptionsList x, int getNumberOfAsmOptions
where
  toBeTested(x) and
  not x.isUnknown() and
  getNumberOfAsmOptions = x.getNumberOfAsmOptions()
select x, "getNumberOfAsmOptions:", getNumberOfAsmOptions
