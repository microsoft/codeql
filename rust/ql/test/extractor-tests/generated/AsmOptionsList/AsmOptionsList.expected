| gen_asm_options_list.rs:8:14:8:36 | AsmOptionsList | getNumberOfAsmOptions: | 2 |
